File: cong-cu/dung-tulien.py
# -*- coding: utf-8 -*-
"""Bộ đồng nghĩa / trái nghĩa tiếng Nhật, XẾP HẠNG theo số synset ủng hộ.

Một từ tiếng Nhật ứng với nhiều synset (nhiều nghĩa) của WordNet. Gom hết ứng
viên của mọi synset rồi cắt lấy 6 cái đầu thì thứ tự là ngẫu nhiên: 始まる mất
終わる mà lại giữ 立休らう. Đếm xem mỗi ứng viên được BAO NHIÊU synset ủng hộ rồi
xếp theo đó thì nghĩa trung tâm nổi lên trước.
"""
import sqlite3, json, re, sys, os, collections

POS_TEP = {"n": "data.noun", "v": "data.verb", "a": "data.adj", "r": "data.adv"}
c = sqlite3.connect("wnjpn.db")

jp = collections.defaultdict(list)
for syn, lemma in c.execute(
        "SELECT sense.synset, word.lemma FROM sense JOIN word USING (wordid) WHERE word.lang='jpn'"):
    jp[syn].append(lemma)

ant = collections.defaultdict(set)
for pos, tep in POS_TEP.items():
    for dong in open("wordnet/" + tep, encoding="utf-8", errors="replace"):
        if dong.startswith("  "): continue
        phan = dong.split("|")[0].split()
        try:
            off, ss_type = phan[0], phan[2]
            i = 4 + int(phan[3], 16) * 2
            p_cnt = int(phan[i]); i += 1
            for _ in range(p_cnt):
                sym, o2, p2 = phan[i], phan[i+1], phan[i+2]; i += 4
                if sym == "!":
                    a = off + "-" + ("a" if ss_type == "s" else ss_type)
                    b = o2 + "-" + ("a" if p2 == "s" else p2)
                    ant[a].add(b); ant[b].add(a)
        except (IndexError, ValueError): continue

def sach(t):
    return bool(t) and len(t) <= 14 and not re.search(r"[.。！？!?、,;:…\n]", t)

tu_syn = collections.defaultdict(set)
for syn, ds in jp.items():
    for t in ds: tu_syn[t].add(syn)

# Một từ nằm trong càng NHIỀU synset thì càng chung chung; dùng để phá hoà.
pho_bien = {t: len(s) for t, s in tu_syn.items()}

TOI_DA = 6
ra = {}
for t, dsSyn in tu_syn.items():
    if not sach(t): continue
    dD, dT = collections.Counter(), collections.Counter()
    for s in dsSyn:
        for w in jp.get(s, ()):
            if w != t and sach(w): dD[w] += 1
        for s2 in ant.get(s, ()):
            for w in jp.get(s2, ()):
                if w != t and sach(w): dT[w] += 1
    xep = lambda d: [w for w, _ in sorted(d.items(), key=lambda kv: (-kv[1], -pho_bien.get(kv[0], 0), len(kv[0])))][:TOI_DA]
    dong, trai = xep(dD), xep(dT)
    if len(dong) >= 2 or len(trai) >= 1:
        o = {}
        if dong: o["dong"] = dong
        if trai: o["trai"] = trai
        ra[t] = o

# Cắt thành 32 mảnh theo mã của chữ đầu (~116 KB mỗi mảnh). Xem chú thích
# napBo() trong tu-lien.js về việc vì sao không nạp cả cục.
SO_MANH = 32
manh = collections.defaultdict(list)
for t, v in ra.items():
    manh[ord(t[0]) % SO_MANH].append(
        t + "\t" + ",".join(v.get("dong", [])) + "\t" + ",".join(v.get("trai", [])))
tong = 0
for thumuc in ("extension/tu-lien", "android/www/tu-lien"):
    os.makedirs(thumuc, exist_ok=True)
    for i in range(SO_MANH):
        d = "\n".join(manh.get(i, []))
        open(os.path.join(thumuc, "%d.txt" % i), "w", encoding="utf-8").write(d)
        if thumuc.startswith("extension"): tong += len(d.encode())
txt = "x" * tong
print("từ:", len(ra), "· có trái nghĩa:", sum(1 for v in ra.values() if "trai" in v),
      "·", round(tong/1048576, 2), "MB ·", SO_MANH, "mảnh", file=sys.stderr)
for t in ["改善", "大きい", "始まる", "難しい", "勉強", "多い", "上がる", "安全"]:
    if t in ra:
        print(" ", t, "→ đồng:", "/".join(ra[t].get("dong", [])[:4]),
              "· trái:", "/".join(ra[t].get("trai", [])[:4]), file=sys.stderr)

File: cong-cu/dung-tulien.mjs
/**
 * Dựng `tu-lien-bo.js` từ 日本語WordNet, CHỈ cho những từ có trong sổ tay.
 *
 *   node cong-cu/dung-tulien.mjs wnjpn.db so-tay.json > extension/tu-lien-bo.js
 *
 * Vì sao cắt theo sổ tay: trọn bộ nặng ~90MB. Nhét vào extension thì phình gói
 * cài và chậm lúc nạp, mà phần lớn số từ trong đó người dùng không học. Vài
 * trăm từ thì tệp ra chỉ vài chục KB.
 *
 * Cần `sqlite3` trên máy (macOS và phần lớn bản Linux đã có sẵn).
 */
import { execFileSync } from "node:child_process";
import { readFileSync } from "node:fs";

const [db, soTay] = process.argv.slice(2);
if (!db || !soTay) {
  console.error("Dùng: node cong-cu/dung-tulien.mjs wnjpn.db so-tay.json > extension/tu-lien-bo.js");
  process.exit(2);
}

/** Những từ tiếng Nhật đang có trong sổ tay. */
const nb = JSON.parse(readFileSync(soTay, "utf8"));
const kho = nb.notebook || nb;
const tu = new Set();
for (const k of Object.keys(kho)) {
  const e = kho[k];
  if (!e || e.del || !e.word) continue;
  if (String(k).indexOf("javi:") === 0 || String(e.dict || "").indexOf("ja") === 0) tu.add(e.word);
}
if (!tu.size) { console.error("Sổ tay không có từ tiếng Nhật nào."); process.exit(1); }

const hoi = (sql) => execFileSync("sqlite3", ["-separator", "\t", db, sql], { encoding: "utf8", maxBuffer: 1 << 28 });

// Từ -> các synset chứa nó
const dsTu = [...tu].map((t) => "'" + t.replace(/'/g, "''") + "'").join(",");
const bang = {};
for (const dong of hoi(`SELECT lemma, synset FROM word JOIN sense USING (wordid) WHERE lemma IN (${dsTu});`).split("\n")) {
  const [lemma, syn] = dong.split("\t");
  if (!lemma || !syn) continue;
  (bang[lemma] = bang[lemma] || { syn: new Set(), dong: new Set(), trai: new Set() }).syn.add(syn);
}

/** Các lemma tiếng Nhật của một synset. */
function lemmaCua(dsSyn) {
  const ra = {};
  if (!dsSyn.length) return ra;
  const l = dsSyn.map((s) => "'" + s + "'").join(",");
  for (const dong of hoi(`SELECT sense.synset, word.lemma FROM sense JOIN word USING (wordid)
      WHERE sense.synset IN (${l}) AND word.lang='jpn';`).split("\n")) {
    const [s, w] = dong.split("\t");
    if (!s || !w) continue;
    (ra[s] = ra[s] || []).push(w);
  }
  return ra;
}

const moiSyn = [...new Set(Object.values(bang).flatMap((x) => [...x.syn]))];
const theoSyn = lemmaCua(moiSyn);

// Trái nghĩa: quan hệ 'ant' nằm ở mức SYNSET trong wnjpn
const trai = {};
if (moiSyn.length) {
  const l = moiSyn.map((s) => "'" + s + "'").join(",");
  for (const dong of hoi(`SELECT synset1, synset2 FROM synlink WHERE link='ant' AND synset1 IN (${l});`).split("\n")) {
    const [a, b] = dong.split("\t");
    if (a && b) (trai[a] = trai[a] || []).push(b);
  }
}
const theoSynTrai = lemmaCua([...new Set(Object.values(trai).flat())]);

const ra = {};
for (const t of Object.keys(bang)) {
  const d = new Set(), tr = new Set();
  for (const s of bang[t].syn) {
    for (const w of theoSyn[s] || []) if (w !== t) d.add(w);
    for (const s2 of trai[s] || []) for (const w of theoSynTrai[s2] || []) tr.add(w);
  }
  const dong = [...d].slice(0, 8), nghich = [...tr].slice(0, 8);
  if (dong.length || nghich.length) ra[t] = { dong, trai: nghich };
}

process.stderr.write("Dựng được " + Object.keys(ra).length + "/" + tu.size + " từ.\n");
process.stdout.write(
  "/* Sinh bằng cong-cu/dung-tulien.mjs từ 日本語WordNet. Đừng sửa tay. */\n" +
  "(function (g) { g.TuLienBo = " + JSON.stringify(ra) + "; })" +
  "(typeof self !== \"undefined\" ? self : this);\n");
